### DISCUSSION TITLE
Use the FileSystemProvider to provide a "virtual" import in typescript

### DISCUSSION THREAD

**Discussion by DavidBal** [microsoft/vscode-discussions#1220] [Extension Development QnA] [answered]
Hi,

i want to provide a "virtual" import to a TypeScript file and get Intellisense for it.

My TypeScript File should look something like this:

![image](https://github.com/microsoft/vscode-discussions/assets/19770870/576ef058-44e8-44d8-98b2-9f63fdb526a8)

```ts
import { Test } from 'virtualimport:/Test.ts';

Test();
```

The FileSystemProvider is registered like this:
```ts
const disposable = vscode.workspace.registerFileSystemProvider("virtualimport", new VirtualImportFileSystemProvider());
```

If the file link (virtualimport:/Test.ts) is clicked with strg the file is opened as expected.

Is there a way to provide this "virtual" file to TypeScript for Intellisense?

Thanks in advance.

Best regards,
David

**Comment by isidorn** [comment]
I do not know.
Do you still need help with this? If yes, I can involve the right folk.

**Reply by DavidBal** [reply]
Hi, thanks for your answer.
Yes, i still need help with this problem.

Regards,
David

**Reply by isidorn** [reply]
Maybe @mjbvz could help. Since I do not know.

**Comment by mjbvz** [comment, marked as answer]
A few issues with this idea:

- On desktop, TypeScript can't read virtual file systems directly. We're investigating support for this but it's challenging
- I'm doubtful that TS can correctly resolve that path syntax 

You can provide virtual files to TypeScript directly using a TypeScript service plugin however. These can be contributed by a VS Code extension too

**Reply by DavidBal** [reply]
Thank you for your answer.

Do you have any documentation/example how to implement this with a ts Service Plugin?

I all ready tried that too but failed...
I did not find the function i need to override/proxy to archiv my goal.

Regards,
David
